generator/internal/printer.h

```diff
@@ -63,6 +65,38 @@ class Printer {
     printer_->Print(text.c_str(), std::forward<Args>(args)...);
   }
 
+  /**
+   * Like Print(), except it accepts diagnostic information from the caller
+   * and rethrows exceptions thrown by protoc in order to provide a more
+   * meaningful diagnostic.
+   */
+  void Print(int line, char const* file,
+             const std::map<std::string, std::string>& variables,
+             std::string const& text) {
+    try {
```

Comment:
Does this even compile with exceptions disabled?  For that matter, do we care?  If protobuf throws to signal errors it seems that this code *requires* exceptions, so we should simply not compile the whole directory on those platforms.

Look at what we do for the `samples` directory in `spanner` and `pubsub` which also require exceptions to compile.